lib/reporting/reports/sales_tax/sales_tax_totals_by_order.rb

```diff
@@ -0,0 +1,175 @@
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/ClassLength
+module Reporting
+  module Reports
+    module SalesTax
+      class SalesTaxTotalsByOrder < Base
+        def search
+          report_line_items.orders
+        end
+
+        def order_permissions
+          return @order_permissions unless @order_permissions.nil?
+
+          @order_permissions = ::Permissions::Order.new(user, ransack_params)
+        end
+
+        def report_line_items
+          # Needed to filter by supplier_id
+          @report_line_items ||= Reporting::LineItems.new(order_permissions, params)
+        end
+
+        def query_result
+          # We'll group the line items by
+          # [tax_rate_id, order_id]
+          # We'll do it in two steps
+          # 1. we group by [[tax_rate], order_id]
+          # 2. we split each key that has more than tax_rate
+          report_line_items.list.group_by do |line_item|
+            [
+              line_item.tax_rates.map(&:id),
+              line_item.supplier_id,
+              line_item.order_id,
+            ]
+          end.flat_map do |k, v|
+            k.first.map do |tax_rate_id|
+              {
+                [tax_rate_id] + k[1..] => v
+              }
+            end
+          end
+        end
```

Comment:
I have to admit I'm not able to follow this one. Maybe it's because it's the end of my day.
I can't see much mention of suppliers in the issue, PR or [this example](https://docs.google.com/spreadsheets/d/1VmcaV2GFVgw-4fysnJn9QZeGbZqPOqfXJJ-GOVVbsTI/edit#gid=877250263).

From what I can understand, this grouping is an extra level of grouping that occurs before the grouping defined in `rules`.  Is this defined here because it was too complex for that section?

Reply:
Yes, that's correct.

Every item returned by `query_result` will:
1. match one row in the final report that you'll see on the HTML page (if we don't count the summary rows)
2. be sent to the functions written on `columns`.

An item will look like

```ruby
{
    [tax_rate_id, supplier_id, order_id] => order 
}
```
I needed that extra level of grouping because one order can much multiple : 
1. tax rates.
2. supplier

when one of the columns' functions is called, we need to know which of the tax rates is targeted in the report row. This can only be determined from the first hash key. 

Now that I think about it again, I can simplify the code by grouping the order instead of line items.